src/state_point.cpp

```diff
@@ -577,58 +565,59 @@ void write_source_point(const char* filename, bool surf_source_bank)
   bool parallel = false;
 #endif
 
-  std::string filename_;
-  if (filename) {
-    filename_ = filename;
-  } else {
-    // Determine width for zero padding
-    int w = std::to_string(settings::n_max_batches).size();
+  if (!filename)
+    fatal_error("write_source_point filename needs a nonempty name.");
 
-    filename_ = fmt::format("{0}source.{1:0{2}}.h5", settings::path_output,
-      simulation::current_batch, w);
+  std::string filename_(filename);
+  const auto extension = get_file_extension(filename_);
+  if (extension == "") {
+    filename_.append(".h5");
+  } else if (extension != "h5") {
+    fatal_error("openmc_source_point was passed an incorrect file "
+                "extension. Must either have no file extension or .h5");
   }
 
   hid_t file_id;
   if (mpi::master || parallel) {
-    file_id = file_open(filename_, 'w', true);
+    file_id = file_open(filename_.c_str(), 'w', true);
     write_attribute(file_id, "filetype", "source");
   }
 
   // Get pointer to source bank and write to file
-  write_source_bank(file_id, surf_source_bank);
+  write_source_bank(file_id, source_bank, bank_index);
 
   if (mpi::master || parallel)
     file_close(file_id);
 }
 
-void write_source_bank(hid_t group_id, bool surf_source_bank)
+void write_source_bank(hid_t group_id, gsl::span<SourceSite> source_bank,
+  vector<int64_t> const& bank_index)
 {
   hid_t banktype = h5banktype();
 
   // Set total and individual process dataspace sizes for source bank
-  int64_t dims_size = settings::n_particles;
-  int64_t count_size = simulation::work_per_rank;
-
-  // Set vectors for source bank and starting bank index of each process
-  vector<int64_t>* bank_index = &simulation::work_index;
-  vector<SourceSite>* source_bank = &simulation::source_bank;
-  vector<int64_t> surf_source_index_vector;
-  vector<SourceSite> surf_source_bank_vector;
-
-  // Reset dataspace sizes and vectors for surface source bank
-  if (surf_source_bank) {
-    surf_source_index_vector = calculate_surf_source_size();
-    dims_size = surf_source_index_vector[mpi::n_procs];
-    count_size = simulation::surf_source_bank.size();
-
-    bank_index = &surf_source_index_vector;
-
-    // Copy data in a SharedArray into a vector.
-    surf_source_bank_vector.resize(count_size);
-    surf_source_bank_vector.assign(simulation::surf_source_bank.data(),
-      simulation::surf_source_bank.data() + count_size);
-    source_bank = &surf_source_bank_vector;
-  }
+  // TODO: are these correct?
+  // the old code was
+  //
+  // int64_t dims_size = settings::n_particles;
+  // int64_t count_size = simulation::work_per_rank;
+  //
+  // but that doesn't make sense! The count on a rank is not
+  // necessarily equal to work_per_rank. For the case of the
+  // fission source bank, that's true, but for surface source
+  // creation, there may be a different number of particles
+  // on each rank.
```

Comment:
If you look at the develop branch, although `count_size` starts off as being equal to `work_per_rank`, for the surface source case it gets set again to `surf_source_bank.size()`:
https://github.com/openmc-dev/openmc/blob/ae615c02d9bb97f05f52a6b74494f1dc37514186/src/state_point.cpp#L619-L622